src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.cs

```diff
@@ -63,6 +63,13 @@ internal partial class AnalysisState
         private readonly ObjectPool<HashSet<CompilationEvent>> _compilationEventsPool;
         private readonly HashSet<CompilationEvent> _pooledEventsWithAnyActionsSet;
 
+        // Create static pools for heavily allocated per-analyzer state objects - this helps in reducing allocations across CompilationWithAnalyzer instances.
+        private const int PoolSize = 1000;
+        private static readonly ObjectPool<AnalyzerStateData> s_analyzerStateDataPool = new ObjectPool<AnalyzerStateData>(() => new AnalyzerStateData(), PoolSize);
```

Comment:
I see your point - and in fact I did few trial runs and found that if the pool size is too small (say 100) then the allocation regression for our perf test still shows. If the size is 1000, we get half the allocations back. If I bump it to 5000, we get more than 80% of the additional allocations back, but there is a noticeable VM increase.

I agree that we do need some cache shrinking or dynamic pool size logic. Let me try to put 3000 item limits for dictionary before dropping it and see the numbers.